variable.c

```diff
@@ -3834,30 +3835,65 @@ rb_iv_set(VALUE obj, const char *name, VALUE val)
 int
 rb_class_ivar_set(VALUE obj, ID key, VALUE value)
 {
-    if (!RCLASS_IV_TBL(obj)) {
-        RCLASS_IV_TBL(obj) = st_init_numtable();
+    RUBY_ASSERT(RB_TYPE_P(obj, T_CLASS) || RB_TYPE_P(obj, T_MODULE));
+    int found;
+
+    RB_VM_LOCK_ENTER();
+    {
+        rb_shape_t * shape = rb_shape_get_shape(obj);
+        attr_index_t idx;
+        found = rb_shape_get_iv_index(shape, key, &idx);
+
+        if (found) {
+            // Changing an existing instance variable
+            RUBY_ASSERT(RCLASS_IVPTR(obj));
+
+            RCLASS_IVPTR(obj)[idx] = value;
+            RB_OBJ_WRITTEN(obj, Qundef, value);
+        } else {
+            // Creating and setting a new instance variable
+
+            // Move to a shape which fits the new ivar
+            shape = rb_shape_get_next(shape, obj, key);
+            RUBY_ASSERT(shape->next_iv_index > 0);
+            idx = shape->next_iv_index - 1;
```

Comment:
A bit of a nitpick, but this might read cleaner to me if we switch the ordering a little and get the `next_iv_index` from the previous shape and then get the new shape when we set it. Something like:

```
idx = shape->next_iv_index;
RUBY_ASSERT(idx >= 0);

....

rb_shape_t * new_shape = rb_shape_get_next(shape, obj, key);
rb_shape_set_shape(obj, new_shape);
```

Reply:
good call. done